[Worxlandroid] Initial contribution

clinique opened the pull request:
This binding is used since a [while](https://community.openhab.org/t/worx-landroid-binding/95246/609) as a standalone - let's get it merged. 
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unusable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

All PRs should be created using the "main" branch as base.
Important bugfixes are cherry-picked by maintainers to the patch release branch after a PR has been merged.

Add one or more appropriate labels to make your PR show up in the release notes.
E.g. enhancement, bug, documentation, new binding
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- [homematic] Improve communication with weak signal devices
- [timemachine][WIP] Initial contribution

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

# Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

It is a good practice to add a URL to your built JAR in this pull request description,
so it is easier for the community to test your Add-on.
If your pull request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Marketplace:
https://community.openhab.org/c/marketplace/69

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

clinique committed:
Worxlandroid new binding

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

clinique committed:
Applying spotless

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

clinique committed:
Code cleansing started

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

wborn commented on bundles/org.openhab.binding.worxlandroid/pom.xml:
```diff
@@ -0,0 +1,30 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.2.0-SNAPSHOT</version>
```
```suggestion
    <version>4.3.0-SNAPSHOT</version>
```

clinique replied:
Done

clinique committed:
Updated to 4.3 and two SAT corrections
Correcting pom.xml

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

clinique committed:
Adding https://github.com/clinique/openhab-addons/pull/24

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+
+# WorxLandroid Binding
+
+This is the binding for Worx Landroid robotic lawn mowers. It connects openHAB with your WorxLandroid mower using the API and MQTT. This binding allows you to integrate, view and control supported Worx lawn mowers in the openHAB environment.
```
```suggestion
This is the binding for Worx Landroid robotic lawn mowers.
It connects openHAB with your WorxLandroid mower using the API and MQTT.
This binding allows you to integrate, view and control supported Worx lawn mowers in the openHAB environment.
```

clinique replied:
Changed

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+
+Currently following Things are supported:
+
+- **Bridge Worx Landroid API** Thing representing the handler for Worx API
+- One or many Things for supported **Landroid Mower**'s
```
can you add the thing id `bridge` and `mower` at the start of the lines?

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+
+## Discovery
+
+A Bridge is required to connect to the Worx API. Here you can provide your credentials for your WorxLandroid account. Once the Bridge has been added Worx Landroid mowers will be discovered automatically.
```
```suggestion
A Bridge is required to connect to the Worx API.
Here you can provide your credentials for your WorxLandroid account.
Once the Bridge has been added Worx Landroid mowers will be discovered automatically.
```

clinique replied:
Changed

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+
+## Binding Configuration
+
+Following options can be set for the **Bridge Worx Landroid API**:
```
The following options...

Can you also add the thing id `bridge` at the end?

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+| password | Password to access the WorxLandroid API. |
+
+
+Following options can be set for the **Landroid Mower**:
```
```suggestion
The following options can be set for the `mower`:
```

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+
+A Bridge is required to connect to the Worx API. Here you can provide your credentials for your WorxLandroid account. Once the Bridge has been added Worx Landroid mowers will be discovered automatically.
+
+## Binding Configuration
```
The configuration below does not seem to be binding configuration, but rrather Thing/Bridge configuration.

clinique replied:
Corrected

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+| Property              | Description                                                                                             |
+|-----------------------|---------------------------------------------------------------------------------------------------------|
+| serialNumber          | Serial number of the mower                                                                              |
+| refreshStatusInterval | Interval for refreshing mower status (ONLINE/OFFLINE) and channel 'common#online' in seconds (min="30") |
+| pollingInterval       | Interval for polling in seconds (min="30" max="7200").                                                  |
```
Did you consider marking these as advanced?

clinique replied:
Sure.

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/landroid_error_de.map:
```diff
@@ -0,0 +1,20 @@
+UNKNOWN=unknown
```
file ends with `de` suggesting it is a german map (or am i mistaken?)
Better to change it to `en`

Same for the order map file.

clinique replied:
Well spotted. Corrected

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+|-----------------------|---------------------------------------------------------------------------------------------------------|
+| serialNumber          | Serial number of the mower                                                                              |
+| refreshStatusInterval | Interval for refreshing mower status (ONLINE/OFFLINE) and channel 'common#online' in seconds (min="30") |
+| pollingInterval       | Interval for polling in seconds (min="30" max="7200").                                                  |
```
the mower.xml thing strucutre has min="0", this seems inconsistent, not sure what it should be.

clinique replied:
Both min are 0, no added value here. Removed.

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This sample file can be removed.

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+| connected | Switch | aws#connected |
+
+
+##### sunday (Slot 1)
```
This is quite extensive. If you add some lines that each day of the week has a slot 1 and 2 and provide a sample for just one day, it would be enough good enough to understand.

clinique replied:
Changed

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
````diff
@@ -0,0 +1,649 @@
+
+### $OPENHAB_CONF/sitemaps/landroid.sitemap
+
+```
````
```suggestion
```perl
```

clinique replied:
Changed

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
````diff
@@ -0,0 +1,649 @@
+
+### $OPENHAB_CONF/items/landroid.items
+
+```
````
```suggestion
```java
```

clinique replied:
Changed

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+
+
+## Examples
+
```
Please add the bridge/mower thing declaration too.

clinique replied:
Added

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,649 @@
+<img src="images/landroid.svg">
```
Please remove, this logo can be uploaded after the binding is merged, so that it ends up in a proper place the UI expects it.

clinique replied:
Changed

lsiepel reviewed (changes requested):
Thanks for contributing this binding. This is a partly review, i only looked at the documentation, thing structure and some metadata files. I'll continue my review when i find some more time.
I noticed an images folder, i guess that can be removed as they do not seem to be used anywhere.

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/thing/mower.xml:
```diff
@@ -0,0 +1,97 @@
+			<parameter name="pollingInterval" unit="s" type="integer" min="0" max="7200">
+				<label>Polling Interval</label>
+				<description>Interval for polling in seconds</description>
+				<default>0</default>
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/thing/mower.xml:
```diff
@@ -0,0 +1,97 @@
+				<required>true</required>
+			</parameter>
+
+			<parameter name="refreshStatusInterval" unit="s" type="integer" min="0">
```
You're correct - this is not appropriate at all.

clinique commented on bundles/org.openhab.binding.worxlandroid/doc/products.json:
```diff
@@ -0,0 +1,1690 @@
+ [
```
None

clinique committed:
Code review n°1 corrections

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,539 @@
+
+# WorxLandroid Binding
+
+This is the binding for Worx Landroid robotic lawn mowers. 
```
```suggestion
This is the binding for Worx Landroid robotic lawn mowers.
```

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,539 @@
+
+# WorxLandroid Binding
+
+This is the binding for Worx Landroid robotic lawn mowers. 
+It connects openHAB with your WorxLandroid mower using the API and MQTT. 
+This binding allows you to integrate, view and control supported Worx lawn mowers in the openHAB environment.
```
```suggestion
This is the binding for Worx Landroid robotic lawn mowers.
It connects openHAB with your WorxLandroid mower using the API and MQTT.
This binding allows you to integrate, view and control supported Worx lawn mowers in the openHAB environment.
```

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,539 @@
+
+## Discovery
+
+A Bridge is required to connect to the Worx API. 
+Here you can provide your credentials for your WorxLandroid account. 
```
```suggestion
A Bridge is required to connect to the Worx API.
Here you can provide your credentials for your WorxLandroid account.
```

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,539 @@
+| username | Username to access the WorxLandroid API. |
+| password | Password to access the WorxLandroid API. |
+
+
```
```suggestion
```

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,539 @@
+
+Currently following **Channels** are supported on the **Landroid Mower**:
+
+##### common
```
Please adjust headers (capitalizxed and level)
```suggestion
### Common
```

clinique replied:
Corrected

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,539 @@
+| allocation-8 | Number        | multi-zones#allocation-8 |
+| allocation-9 | Number        | multi-zones#allocation-9 |
+
+
```
```suggestion
```

clinique replied:
Changed

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,539 @@
+| next-start     | DateTime | schedule#next-start     |                   |
+| next-stop      | DateTime | schedule#next-stop      |                   |
+
+
```
```suggestion
```

clinique replied:
Changed this one and the others

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
```diff
@@ -0,0 +1,539 @@
+| duration | Number:Time | sunday#duration |
+| edgecut  | Switch      | sunday#edgecut  |
+
+
```
Please adjust empty lines in this file
```suggestion
```

clinique replied:
Here also

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
````diff
@@ -0,0 +1,539 @@
+
+### $OPENHAB_CONF/items/landroid.things
+
+```xtend
````
```suggestion
```java
```

clinique replied:
Corrected

lsiepel commented on bundles/org.openhab.binding.worxlandroid/README.md:
````diff
@@ -0,0 +1,539 @@
+
+### $OPENHAB_CONF/rules/landroid.rules
+
+```
````
Add code fence language

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/discovery/MowerDiscoveryService.java:
```diff
@@ -0,0 +1,84 @@
+                thingDiscovered(discoveryResult);
+                logger.debug("Discovered a mower thing with ID '{}'", mower.serialNumber);
+            });
+        } catch (WebApiException npe) {
```
To prevend some future mistake, this is a generic exception and not (only) an npe.

```suggestion
        } catch (WebApiException exception) {
```

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/discovery/MowerDiscoveryService.java:
```diff
@@ -0,0 +1,84 @@
+                logger.debug("Discovered a mower thing with ID '{}'", mower.serialNumber);
+            });
+        } catch (WebApiException npe) {
+            logger.error("Error in WebApiException : {}", npe.getMessage());
```
```suggestion
            logger.warn("Error in WebApiException : {}", npe.getMessage());
```

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/api/WorxApiHandler.java:
```diff
@@ -0,0 +1,127 @@
+    private Request buildRequest(String url, String accessToken, HttpMethod method) {
+        Request request = httpClient.newRequest(url).method(method);
+        request.header(HttpHeader.AUTHORIZATION, "Bearer %s".formatted(accessToken));
+        request.header(HttpHeader.CONTENT_TYPE, "application/json; utf-8");
+        return request;
```
I'd opt to set an explicit timeout (always)
```suggestion
        request.header(HttpHeader.CONTENT_TYPE, "application/json; utf-8");
        request.timeout(15, TimeUnit.SECONDS);
```

clinique replied:
Done

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/api/WorxApiHandler.java:
```diff
@@ -0,0 +1,127 @@
+            ContentResponse response = request.send();
+            if (response.getStatus() == 200) {
+                String result = response.getContentAsString();
+                logger.debug("Worx Landroid Api Response: {}", result);
```
verbose content should go to trace. url metadata can go to debug.
```suggestion
                logger.trace("Worx Landroid Api Response: {}", result);
```

clinique replied:
Changed

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidBridgeHandler.java:
```diff
@@ -0,0 +1,210 @@
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "@text/conf-error-no-password");
+            return;
+        }
+
```
```suggestion
        updateStatus(ThingStatus.UNKNOWN);
```

clinique replied:
Well spotted

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidBridgeHandler.java:
```diff
@@ -0,0 +1,210 @@
+            accessToken = oAuthClientService.getAccessTokenByResourceOwnerPasswordCredentials(username, password, "*")
+                    .getAccessToken();
+
+            if (firstLaunch()) {
```
With this implementation, the properties are never updated after the first launch.
You could double check, but if i remember correctly the properties will be empty each start for file based things and managed things will remember these properties.

clinique replied:
You're right. Here are the properties:
![image](https://github.com/user-attachments/assets/c3316ea9-ae52-4f76-b90c-4f5dfae9cd0e)

Most are not subject to change, but e.g. firmware can change. I remember having had the same discussion on another binding, and the added value of checking first run is poor. I remove it.

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidMowerHandler.java:
```diff
@@ -0,0 +1,577 @@
+                startScheduledJobs(bridgeHandler, mower.get(), config);
+            }
+        } catch (WebApiException e) {
+            logger.error("initialize mower: id {} - {}::{}", config.serialNumber, getThing().getLabel(),
```
```suggestion
            logger.warn("initialize mower: id {} - {}::{}", config.serialNumber, getThing().getLabel(),
```

clinique replied:
Modified

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidMowerHandler.java:
```diff
@@ -0,0 +1,577 @@
+            if (product != null) {
+                connectAws(product.mqttEndpoint, product.uuid, product.userId, product.mqttTopics.commandOut);
+                mower = Optional.of(new Mower(this, product));
+                if (firstLaunch()) {
+                    setChannelsAndProperties(mower.get());
```
See earlier comment about properties.

clinique replied:
Removed here also

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidMowerHandler.java:
```diff
@@ -0,0 +1,577 @@
+            }
+        });
+
+        // What is this ???
```
You tell me ;-)

clinique replied:
Oops, forgotten comment.

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/AWSClientThingHandler.java:
```diff
@@ -0,0 +1,231 @@
+        }
+        WorxLandroidBridgeHandler bridgeHandler = getBridgeHandler(getBridge(), WorxLandroidBridgeHandler.class);
+        if (bridgeHandler == null) {
+            logger.error("no bridgeHandler");
```
```suggestion
            logger.warn("no bridgeHandler");
```

clinique replied:
Done

clinique committed:
Code review n°2 corrections

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

lsiepel commented on bundles/org.openhab.binding.worxlandroid/pom.xml:
```diff
@@ -0,0 +1,30 @@
+
+  <dependencies>
+    <dependency>
+      <groupId>software.amazon.awssdk.iotdevicesdk</groupId>
+      <artifactId>aws-iot-device-sdk</artifactId>
+      <version>1.15.0</version>
+    </dependency>
+    <dependency>
+      <groupId>software.amazon.awssdk.crt</groupId>
+      <artifactId>aws-crt</artifactId>
+      <version>0.25.1</version>
+    </dependency>
+  </dependencies>
```
This binding performs some similar requests/response, so it should be possible to create them without these SDK's.
Could you remove these dependencies?

clinique replied:
Can you be more explicit ? I'm afraid I did not get the meaning of your comment.
IIRW I tried to upgrade dependencies but it had major impact on code structure, so prefered to leave as it was.

lsiepel replied:
I meant to say that the SDK should not be used. Instead the httpclient can be used to perform the same api request/responses.

You could look at the salus binding that also holds a very light weight base class doing much of the SDK work without needing the actual SDK. 

Due to problems of the past with size, upgrades and other issues we try to keep these SDK’s out of the codebase.

clinique replied:
This is for Mqtt client of AWS IoT

lsiepel replied:
Have not looked in depth on how the MQTT client and websocket / authenticatino work together. I just know that we tent not to accept these SDK's. For the salus binding we had a similar issue with the `software.amazon.awssdk.crt` package. A fix was found by some lightweight classes that implemented the authentication. 
Hope you figure out hwo to get rid of the SDK with these suggestions, otherwise please post the specific problems you encounter and maybe we can find a fix together. If ultimately it seems not feasable, we can always merge it as is, but yeah really prefer not to..

clinique replied:
I tried to incorporate limite set of classes from the SDK but I think I fall in the same problem with the crt package. Will take a look on how it has been handled for Salus.

clinique replied:
@lsiepel : had no success here.

lsiepel replied:
We will need to look for a solution as the jar file seems to explode with this SDK. It grows to a whopping 17MB.

clinique replied:
Using `org.openhab.core.io.transport.mqtt` was my initial intention but as you identified it connection uses a mix of `mqtt` and `websocket` that I've not been able to reproduce in an other environment.

lsiepel replied:
Are you aware of a simulator or demo api? I don’t have such a device or account. To help figure out, this would be useful.

clinique replied:
Not as far as I know. I Can give you m'y credentials by PM if you agree

clinique replied:
@lsiepel : gentle ping.

jlaur replied:
See also https://github.com/openhab/openhab-addons/pull/14669#issuecomment-1489217564.

lsiepel replied:
Have no0t forgotten this, but have not yet had the dedicated time to look at it. Would be nice if we can either build our own or use a lite mqtt over websockets client.

lsiepel replied:
I would like a comment from @openhab/core-maintainers about this SDK. We now have two bindings that want to use this SDK. 
AFAICT there does not seem to be a way to use the default MQTT client. That project does not seem to have active development. So i see three options:
1. Allow this SDK (increase of bundle size, but re-usable between bindings)
2. Not accept this binding contribution (maybe marketplace is an option?)
3. Wait and/or contribute to the default MQTT client. (maintin status quo)

lsiepel replied:
@jlaur any spare time to look at this comment?

clinique replied:
I tried HiveMQ client in my eatly tentative. I do not find back what blocked me at the time but the result is I did not suceed in that way.

lsiepel replied:
> I can't find any problems so far with the URL encoder in the HiveMQ implementation, is that still a problem and why?

Did you read the linked comment? As that also points to: https://github.com/hivemq/hivemq-mqtt-client/issues/643
Maybe @jpg0 can elaborate on the issue and if your suggestion is a workaround.
@clinique would it be possible for you to give it another try and describe where you encounter issues? It would be helpfull to either find a workaround or a decision on SDK usage.

jpg0 replied:
I filed that issue quite a while ago. I have just tried to dig through this as I had originally come to the same conclusion as you, but could not make it work. I've managed to reduce it to the following code:

```
import java.net.URI;

class Main {
    public static void main(String[] args) throws Exception {
        String input = "x/y";
        URI uri = new URI(
            "wss", null, "host", 8080, "/test", input, null);
            
        //Try to output: x%2Fy
        System.out.println(uri.getRawQuery());
    }
}
```

In the above it creates a URI just like hivemq, then passes it on to Netty which extracts the raw query as you pointed out. The challenge is to get the string that Netty uses (and is returned by `getRawQuery()`) to be equal to `x%2Fy` (AWS explicitly checks the exact URI for sig validation, so it cannot be different at all). If you pass it unencoded (x/y) is remains this way, if you pass it encoded (x%2Fy) it double encodes it.

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/i18n/worxlandroid.properties:
```diff
@@ -0,0 +1,355 @@
+channel-group-type.worxlandroid.battery-group-type.label = Battery
+channel-group-type.worxlandroid.battery-group-type.description = Battery channels of your mower
+channel-group-type.worxlandroid.battery-group-type.channel.charge-cycles.label = Current Charge Cycles
+channel-group-type.worxlandroid.battery-group-type.channel.charge-cycles-total.label = Total Charge Cycles
```
=> mvn i18n:generate-default-translations

```suggestion
channel-group-type.worxlandroid.battery-group-type.channel.charge-cycles.label = Charge Cycles
```

clinique replied:
Corrected

lsiepel reviewed (changes requested):
Last bits are reviewed now, covered all files. Mainly documentation and a bigger one regarding the dependencies. It also looks like a netatmo file accidently became part of this PR.

The OP has no contents, could you link a community thread and or somewhere this binding has been tests or discusssed before? I noticed many file are signed by you, and some by 'Nils' did you proceed on someone else's PR?

clinique commented:
> Last bits are reviewed now, covered all files. Mainly documentation and a bigger one regarding the dependencies. It also looks like a netatmo file accidently became part of this PR.
> 
> The OP has no contents, could you link a community thread and or somewhere this binding has been tests or discusssed before? I noticed many file are signed by you, and some by 'Nils' did you proceed on someone else's PR?

Updated OP to the community thread. I took over this binding for the 3.x => 4 migration, despite many requests the original creator ( Nibi79 in the community) never had time to push to common repo, so I decided to proceed. We had a discussion (him and myself) on this. He can take  ownership of the PR as soon as he wants.

clinique committed:
Code review n°3 corrections

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

clinique committed:
Restoring lost Netatmo file

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

jlaur reviewed (commented):
I found a few old pending comments, please find them below.

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/WorxLandroidBindingConstants.java:
```diff
@@ -0,0 +1,110 @@
+ * The {@link WorxLandroidBindingConstants} class defines datCommon constants, which are
+ * used across the whole binding.
+ *
+ * @author Nils - Initial contribution
```
Asked to nibi79 these informations, I'll update accordingly if/when I have the information.

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/thing/groups.xml:
```diff
@@ -0,0 +1,222 @@
+	</channel-group-type>
+
+	<channel-group-type id="wifi-group-type">
+		<label>Wifi Information</label>
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,241 @@
+	</channel-type>
+
+	<channel-type id="rssi" advanced="true">
+		<item-type>Number:Power</item-type>
```
Added

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,241 @@
+		<label>RSSI</label>
+		<description>Received signal strength indicator</description>
+		<category>QualityOfService</category>
+		<state readOnly="true" pattern="%.0f dBm"/>
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,11 @@
+
+	<type>binding</type>
+	<name>WorxLandroid Binding</name>
+	<description>This is the binding for WorxLandroid.</description>
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/images/SC_PaperUI_Mower.png:
These images are no more used in the documentation. It's a relief of v2.x original binding. I removed them.

lsiepel replied:
When reviewing i usually ask the contributor to remove these type of screenshots for the same reason. Better to use links to generic documentation or only allow them when there is a servlet with custom pages but even then I prefer to be hesitant. They are out of date very quickly

clinique committed:
Code review n°4

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

clinique committed:
Adding Nils family name

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

wborn commented on bundles/org.openhab.binding.worxlandroid/pom.xml:
```diff
@@ -0,0 +1,30 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.3.0-SNAPSHOT</version>
```
```suggestion
    <version>5.0.0-SNAPSHOT</version>
```

clinique committed:
Pom update to 5.0
Updated to 4.3 and two SAT corrections
Correcting pom.xml

Signed-off-by: Gaël L'hopital <gael@lhopital.org>

clinique committed:
Update the headers

Signed-off-by: clinique <gael@lhopital.org>

lsiepel commented on bundles/org.openhab.binding.worxlandroid/pom.xml:
```diff
@@ -0,0 +1,30 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>5.0.0-SNAPSHOT</version>
```
```suggestion
    <version>5.1.0-SNAPSHOT</version>
```

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/codes/WorxLandroidStatusCodes.java:
```diff
@@ -0,0 +1,68 @@
+    @SerializedName("34")
+    PAUSE,
+    @SerializedName("99")
+    MANUEL_STOP;
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/vo/Mower.java:
```diff
@@ -0,0 +1,413 @@
+        return result;
+    }
+
+    public Object[] getSheduleArray1() {
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/vo/Mower.java:
```diff
@@ -0,0 +1,413 @@
+        return getScheduleArray(schedules.get(0));
+    }
+
+    public Object[] getSheduleArray2() {
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidMowerHandler.java:
```diff
@@ -0,0 +1,574 @@
+                setScheduledDays(theMower, groupId, channelId, command);
+                sendCommand(theMower,
+                        theMower.scheduler2Supported()
+                                ? new ScheduleDaysCommand(theMower.getTimeExtension(), theMower.getSheduleArray1(),
+                                        theMower.getSheduleArray2())
+                                : new ScheduleDaysCommand(theMower.getTimeExtension(), theMower.getSheduleArray1()));
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidMowerHandler.java:
```diff
@@ -0,0 +1,574 @@
+                sendCommand(theMower,
+                        theMower.scheduler2Supported()
+                                ? new ScheduleDaysCommand(theMower.getTimeExtension(), theMower.getSheduleArray1(),
+                                        theMower.getSheduleArray2())
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidMowerHandler.java:
```diff
@@ -0,0 +1,574 @@
+                setScheduledDays(theMower, groupId, channelId, command);
+                sendCommand(theMower,
+                        theMower.scheduler2Supported()
+                                ? new ScheduleDaysCommand(theMower.getTimeExtension(), theMower.getSheduleArray1(),
+                                        theMower.getSheduleArray2())
+                                : new ScheduleDaysCommand(theMower.getTimeExtension(), theMower.getSheduleArray1()));
```
Corrected

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/WorxLandroidMowerHandler.java:
```diff
@@ -0,0 +1,574 @@
+            sendCommand(theMower,
+                    theMower.scheduler2Supported()
+                            ? new ScheduleDaysCommand(theMower.getTimeExtension(), theMower.getSheduleArray1(),
+                                    theMower.getSheduleArray2())
```
Corrected

clinique committed:
Pom update to 5.1
Copilot review correction
Pom update to 5.0
Updated to 4.3 and two SAT corrections
Correcting pom.xml

Signed-off-by: Gaël L'hopital <gael@lhopital.org>
Signed-off-by: clinique <gael@lhopital.org>

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/resources/OH-INF/thing/groups.xml:
```diff
@@ -0,0 +1,222 @@
+
+	<channel-group-type id="aws-group-type">
+		<label>Aws</label>
+		<description>MQTT connexion to AWS</description>
```
Corrected

clinique committed:
Remove unwanted netatmo file.

Signed-off-by: clinique <gael@lhopital.org>

clinique committed:
Restore Netatmo file

Signed-off-by: clinique <gael@lhopital.org>

clinique committed:
Forgot ressource files

Signed-off-by: clinique <gael@lhopital.org>

lsiepel commented:
@clinique i asked other @openhab/add-ons-maintainers to look at the SDK. I give it some more days and will proceed with the review.

lsiepel commented:
Ok, @psmedley @jpg0 and @clinique if we can re-use this SDK i'm fine by adding with that many bindings to rely on it.
Maybe best to create an osgify bundle ? @wborn any advice?

lsiepel reviewed (approved):
@florian-h05 are you able to dependency files to jfrog?

clinique commented:
> OSGi-ify them in openhab-osgiify and set scope=provided in the POM for the dependencies.

Who is doing the OSGiification ? I'll update the code as soon as done. Thanks for helping me close this PR.

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,9 @@
+
+	<feature name="openhab-binding-worxlandroid" description="Worx Landroid Binding" version="${project.version}">
+		<feature>openhab-runtime-base</feature>
+		<bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.worxlandroid/${project.version}</bundle>
```
Ok,  a bit weird the test succeeded, but there was something wrong with the package. So i adapted and hopefully this is fine. The package needs to be: software.amazon.awssdk.iotdevicesdk.aws-iot-device-sdk (after it is merged)
https://github.com/openhab/openhab-osgiify/pull/91

lsiepel reviewed (approved):
Thanks, LGTM

We need to wait for the osgiified lib to be merged and afterwards the build here should be successfull

clinique committed:
Rebased

Signed-off-by: gael@lhopital.org <gael@lhopital.org>

clinique committed:
Update feature.xml

Signed-off-by: gael@lhopital.org <gael@lhopital.org>

lsiepel reviewed (changes requested):
The rebase also caused this to be missing. I added this locally, but now it claims it needs org.graalvm.sdk.
According to maven, this is an optional module. Do you know if this is essential or not? I can adapt the osgi package. I also wonder what we did wrong with the previous test as it was confirmed, but now we have these issues.

clinique committed:
Redoing pom.xml modifications

Signed-off-by: gael@lhopital.org <gael@lhopital.org>

clinique commented:
> ^ what are your thoughts about graal?

I discover this dependency with your message. Do you think it could have been introduced with the version update ?

lsiepel commented:
> > ^ what are your thoughts about graal?
> 
> I discover this dependency with your message. Do you think it could have been introduced with the version update ?

Yes, it was introduced with version upgrade

clinique commented:
> > > ^ what are your thoughts about graal?
> > 
> > 
> > I discover this dependency with your message. Do you think it could have been introduced with the version update ?
> 
> Yes, it was introduced with version upgrade

Is it worth trying to package it in osgii ?

clinique committed:
Rebased
Update headears
Apply spotless
Correct against SAT

Signed-off-by: gael@lhopital.org <gael@lhopital.org>

clinique committed:
Clean SAT warning
Clean icon set

Signed-off-by: clinique <gael@lhopital.org>

clinique committed:
Revert modification of bundle pom

Signed-off-by: clinique <gael@lhopital.org>

clinique committed:
Move to Instant

Signed-off-by: clinique <gael@lhopital.org>

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/api/WorxApiHandler.java:
```diff
@@ -0,0 +1,129 @@
+            throw new WebApiException(
+                    "Error calling Worx Landroid Api! HTTP Status = %d".formatted(response.getStatus()));
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+            throw new WebApiException(e);
```
Applies at more places in this class

lsiepel reviewed (changes requested):
Some more comments i noticed when reading all PR details again.
In the mean whil i look at the AWS part.

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/api/WorxApiHandler.java:
```diff
@@ -0,0 +1,129 @@
+        logger.debug("URI: {}", request.getURI().toString());
+        try {
+            ContentResponse response = request.send();
+            if (response.getStatus() == 200) {
```
Done

clinique commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/config/MowerConfiguration.java:
```diff
@@ -0,0 +1,36 @@
+ */
+@NonNullByDefault
+public class MowerConfiguration {
+    public static final String SERIAL_NUMBER = "serialNumber";
```
Done

clinique committed:
lsiepel code review

Signed-off-by: Gael L'hopital <gael@lhopital.org>

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/AWSClientThingHandler.java:
```diff
@@ -0,0 +1,236 @@
+    @Override
+    public void onAWSConnectionFailed(@Nullable String message) {
+        updateChannelOnOff(GROUP_AWS, CHANNEL_CONNECTED, awsClient.isConnected());
+        updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "No AWS Connection");
```
What about UNKNOWN ? If the http connection is finished, but MQTT not yet, it is not fully finished yet.

Maybe OFFLINE is a bit too much as it is still in transition to become ONLINE, but of the connection (with all of its layers) is not yet fully established, ONLINE is also not true.
Not sure how the channel comes into play, showing connection state in a channel is not fully inline with the openHAB model.

lsiepel commented on bundles/org.openhab.binding.worxlandroid/src/main/java/org/openhab/binding/worxlandroid/internal/handler/AWSClientThingHandler.java:
```diff
@@ -0,0 +1,236 @@
+        }
+        WorxLandroidBridgeHandler bridgeHandler = getBridgeHandler(getBridge(), WorxLandroidBridgeHandler.class);
+        if (bridgeHandler == null) {
+            logger.error("no bridgeHandler");
```
Yeah, sorry it was a bit cryptic:
Referring to this comment: https://github.com/openhab/openhab-addons/pull/16893/changes#r1748075813
<img width="743" height="451" alt="Image" src="https://github.com/user-attachments/assets/a8e1c243-ec1f-4d41-8b2b-fd695a77a7b2" />

clinique committed:
lsiepel codereview
documentation of provided icons

Signed-off-by: Gael L'hopital <gael@lhopital.org>

lsiepel reviewed (changes requested):
I finally had some time to look into the osgifying the SDK. I tried it before, but at the time i thought the mechanism worked different. 
This is corrected in: https://github.com/openhab/openhab-osgiify/pull/116
When that PR is merged and the suggested changes here are done, i expect this to work as intended.

clinique committed:
Prepare POM to osgification

Signed-off-by: clinique <gael@lhopital.org>

clinique commented:
@lsiepel : pom and feature files modified upon your directions.

lsiepel commented on bundles/org.openhab.binding.worxlandroid/pom.xml:
```diff
@@ -0,0 +1,32 @@
+    <dependency>
+      <groupId>software.amazon.awssdk.iotdevicesdk</groupId>
+      <artifactId>aws-iot-device-sdk</artifactId>
+      <version>1.3.16</version>
```
```suggestion
      <version>1.30.2</version>
```

Don;t know why i suggested that old verison, probably a copy/paste mistake.

clinique replied:
You're welcome !

lsiepel committed:
Apply suggestion from @lsiepel

Signed-off-by: lsiepel <leosiepel@gmail.com>

lsiepel committed:
Fix dependency resolution

Signed-off-by: Leo Siepel <leosiepel@gmail.com>

lsiepel reviewed (approved):
Just pushed a small commit to sort the dependency troubles. It should all be sorted now. @clinique could you verify all works as expected? 

I reviewed the binding. LGTM

Would be nice if @jlaur can review the dependency changes i made and/or trust @clinique test's, so we can merge this.

You could add the addon's logo to the openHAB website. See https://www.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website-and-the-ui

clinique commented:
@lsiepel : unfortunately this does not seem to work at the moment. I updated demo.app, cleaned project, mvn clean install. Same thing for worxlandroid binding, pulling your last update. When I try to resolve dependencies, I get: 
```
Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=org.openhab.binding.worxlandroid)
          ⇒ [org.openhab.binding.worxlandroid version=5.2.0.202604130831]
              ⇒ osgi.wiring.package: (osgi.wiring.package=software.amazon.awssdk.crt)
```

clinique commented:
In the meantime, I'm sure we'll succeed, I created the [PR for the logo](https://github.com/openhab/openhab-docs/pull/2702)

lsiepel commented:
When the jar is dropped in the addon folder, you also need to provide the dependency's that are not bundled in the jar itself. You can download the aws-crt dependency file here:
https://1drv.ms/u/c/70ea7ac46bc61c73/IQCG8KkDQICiTopiSvcTwzxtAbkDcj1XWXjRfiuSuBn1OkU?e=NyMCFg

Shipped bindings  will do this automatically by karaf/felix. While troubleshooting i discovered something was wrong with the osiified library. I'll create a PR right away to fix that, it is already included in the linked file above.
Please continue your test with this file.

clinique commented:
> When the jar is dropped in the addon folder, you also need to provide the dependency's that are not bundled in the jar itself. You can download the aws-crt dependency file here: https://1drv.ms/u/c/70ea7ac46bc61c73/IQCG8KkDQICiTopiSvcTwzxtAbkDcj1XWXjRfiuSuBn1OkU?e=NyMCFg
> 
> Shipped bindings will do this automatically by karaf/felix. While troubleshooting i discovered something was wrong with the osiified library. I'll create a PR right away to fix that, it is already included in the linked file above. Please continue your test with this file.

I did put your provided jar along with the binding jar in /usr/share/openhab/addons and it works:
<img width="613" height="29" alt="image" src="https://github.com/user-attachments/assets/797cf347-3557-4166-a23a-6d7c3275181b" />

Good job @lsiepel

lsiepel commented:
The comment https://github.com/openhab/openhab-addons/pull/16893#pullrequestreview-4093705857 has been burried a little bit, so just for @jlaur a repost. Feel free to merge whenever you are ready.

jlaur commented:
> The comment [#16893 (review)](https://github.com/openhab/openhab-addons/pull/16893#pullrequestreview-4093705857) has been burried a little bit, so just for @jlaur a repost. Feel free to merge whenever you are ready.

Sorry for the slow response, read it on my phone and forgot it a few times, since I needed to look at this proper from a computer. 🙂 If you mean ebdd5ec8a33057b79c007c2527ac5671259073b6, then that LGTM. 👍

lsiepel commented:
Yes, that was what i meant. In the meanwhile the osgified dependency had an optimization that marked the graalvm dependency optional, so i'll remove it here and merge the PR as is.
Give me a day

lsiepel committed:
Remove optional dependency

Signed-off-by: Leo Siepel <leosiepel@gmail.com>
